File: bootstrap/src/main/java/net/dontdrinkandroot/wicket/bootstrap/component/nav/RepeatingNavTabs.java
/*
 * Copyright (C) 2012-2017 Philip Washington Sorst <philip@sorst.net>
 * and individual contributors as indicated
 * by the @authors tag.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.dontdrinkandroot.wicket.bootstrap.component.nav;

import net.dontdrinkandroot.wicket.bootstrap.behavior.NavTabsBehavior;
import org.apache.wicket.model.IModel;

/**
 * @author Philip Washington Sorst <philip@sorst.net>
 */
public class RepeatingNavTabs<T> extends AbstractRepeatingNav<T>
{
    private NavTabsBehavior navTabsBehavior = new NavTabsBehavior();

    public RepeatingNavTabs(String id)
    {
        this(id, null);
    }

    public RepeatingNavTabs(String id, IModel<T> model)
    {
        super(id, model);
        this.add(this.navTabsBehavior);
    }

    public void setJustified(boolean justified)
    {
        this.navTabsBehavior.setJustified(justified);
    }
}

File: bootstrap/src/main/kotlin/net/dontdrinkandroot/wicket/bootstrap/component/nav/RepeatingNavTabs.kt
package net.dontdrinkandroot.wicket.bootstrap.component.nav

import net.dontdrinkandroot.wicket.bootstrap.behavior.NavTabsBehavior
import net.dontdrinkandroot.wicket.bootstrap.component.item.ItemView
import net.dontdrinkandroot.wicket.bootstrap.css.NavItemAlignment
import org.apache.wicket.MarkupContainer
import org.apache.wicket.behavior.Behavior
import org.apache.wicket.model.IModel
import org.apache.wicket.model.Model

abstract class RepeatingNavTabs<T>(
    id: String,
    model: IModel<T>? = null,
    itemAlignmentModel: IModel<NavItemAlignment> = Model(null),
    vararg behaviors: Behavior
) : AbstractRepeatingNav<T>(id, model, *behaviors) {

    init {
        this.add(NavTabsBehavior(itemAlignmentModel))
    }
}

inline fun <T> navTabs(
    id: String,
    model: IModel<T>? = null,
    itemAlignment: IModel<NavItemAlignment> = Model(null),
    vararg behaviors: Behavior,
    crossinline populateItemsHandler: ItemView.(repeatingNavTabs: RepeatingNavTabs<T>) -> Any?
) = object : RepeatingNavTabs<T>(id, model, itemAlignment, *behaviors) {
    override fun populateItems(itemView: ItemView) {
        populateItemsHandler(itemView, this)
    }
}

inline fun navTabs(
    id: String,
    itemAlignment: IModel<NavItemAlignment> = Model(null),
    vararg behaviors: Behavior,
    crossinline populateItemsHandler: ItemView.() -> Any?
) = object : RepeatingNavTabs<Void>(id, null, itemAlignment, *behaviors) {
    override fun populateItems(itemView: ItemView) {
        populateItemsHandler(itemView)
    }
}

inline fun MarkupContainer.addNavTabs(
    id: String,
    itemAlignment: IModel<NavItemAlignment> = Model(null),
    vararg behaviors: Behavior,
    crossinline populateItemsHandler: ItemView.() -> Any?
): RepeatingNavTabs<Void> {
    val repeatingNavTabs = navTabs(id, itemAlignment, behaviors = behaviors, populateItemsHandler)
    add(repeatingNavTabs)
    return repeatingNavTabs
}